src/main/java/org/sagebionetworks/bridge/exporter/worker/TsvInfo.java

```diff
@@ -70,11 +78,31 @@ public void checkInitAndThrow() throws BridgeExporterException {
     public void flushAndCloseWriter() throws BridgeExporterException {
         checkInitAndThrow();
 
-        writer.flush();
-        if (writer.checkError()) {
-            throw new BridgeExporterException("TSV writer has unknown error");
+        // Error handling code here is a bit of a mess. Internally, CSVWriter creates a PrintWriter, which doesn't
+        // throw, but exposes checkError() check for errors. However, CSVWriter declares that flush() throws, even
+        // though internally it doesn't, and then in close(), it calls close on both the PrintWriter and the internal
+        // Writer, one of which closes, and the other does not. Because we need to test with the actual CSVWriter
+        // implementation to make sure all the libraries are doing what we expect, this makes testing error handling
+        // code difficult at best. As such, none of the error handling code is exercised by tests. Fortunately, we log
+        // and rethrow very thoroughly, so if an error occurs, we shouldn't have any problems.
+        try {
+            tsvWriter.flush();
+            if (tsvWriter.checkError()) {
+                LOG.error("TSV writer has unknown error");
+                throw new BridgeExporterException("TSV writer has unknown error");
+            }
+        } catch (IOException ex) {
+            LOG.error("Error flushing TSV writer: " + ex.getMessage(), ex);
+            throw new BridgeExporterException("Error flushing TSV writer: " + ex.getMessage(), ex);
+        } finally {
+            try {
+                tsvWriter.close();
+            } catch (IOException ex) {
+                LOG.error("Error closing TSV writer: " + ex.getMessage(), ex);
+                //noinspection ThrowFromFinallyBlock
+                throw new BridgeExporterException("Error closing TSV writer: " + ex.getMessage(), ex);
+            }
```

Comment:
Would try-with-resources help here? The logging of the IOException would probably not differentiate between an error on flush and an error on close, but that might be in the origin exception message.

Reply:
Try-with-resources doesn't actually work here, because the CSVWriter is declared and created outside of the flushAndCloseWriter() method. I don't believe the code can be easily refactored to use try-with-resources, since the Writer can exist for a very long time and across multiple threads.